src/types/response.ts

```diff
@@ -1,9 +1,28 @@
-export type TypedResponse<data> = {
-  data: data
-  format: 'cast-action' | 'frame' | 'transaction'
-}
+import type { ErrorFrameResponse } from './frame.js'
 
-export type HandlerResponse<typedResponse> =
+export type TypedResponse<
+  data,
+  format extends 'cast-action' | 'frame' | 'transaction',
+> = format extends 'cast-action' | 'transaction'
+  ? {
+      data: data
+      format: format
+    }
+  : { format: format } & (
+      | {
+          isErrorResponse: true
+          error: ErrorFrameResponse
+        }
+      | {
+          isErrorResponse: false
+          data: data
+        }
+    )
```

Comment:
Wonder if we can simplify this: https://www.typescriptlang.org/play?#code/C4TwDgpgBAKuEBMBKEDOYD2A7VEA8CAhsIQDRQQBOlGlUAvFAK5YDWWGA7lgHwNQBvAFBQoAM1oBbYgC4oAcgDGhVMAC0hRcACW2eVAA+CsZUKSI+o-OCmcmnXqEBfKADIoAeSwQPYvAKgiEjkgsihVYiZUOXlUJkVFNFR9FyMAqhpKOQzacgjgKJicyhSeISFFbFUoSjRMHAgAfRMzaEYsJgAbTqgVZjYObl7UWHhkOqr8VUptLABzcgDzVFRCOYg5adm5qCcyoVr0SebTcwA6UKEAeivRO-v7gD0AfgOJhpPWs+Lr24f-qAvcqHeq4T7nfJRBj0RixeKJFb6VzuEHHFrnS43AHYnG4vEAoFvI4fdEQM6QkYw2HFJEo95g0nfai0M7LVbrX74rncnmiQk3AWCoXCkWisXi0XlUCQKAAUQAboROkxiPhpRA+IwAgBtVgQEBQWZQPUgDBiKDqgC6cnVuv1lt2bn67C4WCl8E83l8eBE-V0WAoAA9gBAsAgRhgAEYAKwgWlIvoFvpNIwgwdD4agAGl9WaAKpYf14Fj+zXZ3NiAtFkvYHgJss1gNpkNhkazMRUKAASRDkl9zzliuVqrwPYgkidOqzhoDssDimVCHwKfIJrN3d7PEtzzkLCXYlmiF2ZVEcm88qoQnV5dNlcL2Dw6rL1+bGZG14Ha-N17PEAvlCAA